torch/_prims/nvfuser_prims.py

```diff
@@ -445,12 +448,54 @@ def _prim_impl(
         )
 
     nvprim_impl.impl(name, _prim_impl)
-    nvprim_autograd_impl.impl(
-        name, backwards_not_supported(torch.ops.nvprims.native_batch_norm.default)
-    )
-
     prim_packet = torch.ops.nvprims.native_batch_norm
     prim = prim_packet.default
+
+    def _native_batch_norm_ref(
```

Comment:
What is the output type of `mean` and `rstd`?
Would this work?
```suggestion
    @elementwise_type_promotion_wrapper(
        type_promoting_args=("input",),
        type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT,
    )
    def _native_batch_norm_ref(
```

Reply:
type-promotion wrapper promotes all input and output tensors? So with elementwise promotion, output would also be in promoted math type, which is not what we want.

On paper, the promotion here for BN should only promote `input` and not on other things. e.g. you can't really promote running_stats, because those buffers should be in-place updated. weight/bias as well as mean/rstd should also come in `compute_math`.

Hence I'm only promoting things manually on `input` and cast `output` back to original dtype. BN behavior is strange....

Reply:
`type_promoting_args=("input",),` specifies that only `input` should be promoted.

Reply:
Lol, didn't read that carefully enough.
Does it cast outputs back to the original input dtype? I thought that's only available with reduction promotion.

Reply:
> Does it cast outputs back to the original input type?

Yes, but it might also try to cast all outputs, that's why I'm asking what's the expected type of `mean` and `rstd` outputs.

Reply:
For "cpu", `mean` and `rstd` were cast back to input dtype.
https://github.com/pytorch/pytorch/pull/87116/files#diff-0e8654681a3dcc5fe5ed434f4fc23d2a47d69421eb261371aebc8754d8de5e06R2895-R2898

Reply:
> > Does it cast outputs back to the original input type?
> 
> Yes, but it might also try to cast all outputs, that's why I'm asking what's the expected type of `mean` and `rstd` outputs.

Got it.

Looks like cuda kernel just output mean/var as fp32.
```
In [24]: w = torch.randn(5, device="cuda")

In [25]: b = torch.randn(5, device="cuda")

In [26]: o = torch.ops.aten.native_batch_norm(x, w, b, None, None, True, 0.1, 1e-5)

In [27]: o[0].dtype
Out[27]: torch.float16

In [28]: o[1].dtype
Out[28]: torch.float32

In [29]: o[2].dtype
Out[29]: torch.float32
```

Reply:
Yep, looks like it indeed cast everything back to `input`. so we can't use it here if we want to preserve the behavior from aten.

```
In [33]: @elementwise_type_promotion_wrapper(
    ...:     type_promoting_args=("input",),
    ...:     type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT,
    ...: )
    ...: def func(input, x):
    ...:     print(input.dtype)
    ...:     return (input + x, input, x)
    ...: 

In [34]: o = func(x, w)
torch.float32

In [35]: o[0].dtype
Out[35]: torch.float16

In [36]: o[1].dtype
Out[36]: torch.float16
```

Reply:
tagging @IvanYashchuk to get the approval. I'm leaning towards keeping the code as-is, until later we go fix all the batchnorm consistency issue between cpu/gpu. :vomiting_face: